src/Storages/IStorageCluster.cpp

```diff
@@ -207,6 +209,17 @@ void ReadFromCluster::initializePipeline(QueryPipelineBuilder & pipeline, const
         if (pipes.size() >= max_replicas_to_use)
             break;
 
+        if (skip_non_cache_locality_shards && this->extension->has_pending_task_checker.has_value())
```

Comment:
I think it would be more correct to instead pass a flag `skip_non_cache_locality_shards` to `getNextTask` and just quit right before `getAnyUnprocessedFile` if nothing has been found until that point.

Reply:
Hi @kssenii  thx for review this ! 👍 
IMO if use `getNextTask ` to check if has the locality task is in executor side logic and need create connection in all cluster.
If using the `has_pending_task_checker` it will all do in the scheduler without creating unused connection between scheduler and shard.  what would you think ?